wallet: Remove trailing separators from -walletdir arg

PierreRochard opened the pull request:
If a user passes in a path with a trailing separator as the `walletdir`, multiple BerkeleyEnvironments may be created in the same directory which can lead to data corruption.

Discovered while reviewing https://github.com/bitcoin/bitcoin/pull/12493#issuecomment-417147646

PierreRochard commented:
Would it be better if the test mocked the `g_dbenvs` in db.cpp? I wasn't able to figure out how to do that.

PierreRochard commented:
I'm unable to reproduce the CI failures locally for debugging and the logs don't seem to provide enough information to pinpoint the problem

ken2812221 commented on src/wallet/db.cpp:
```diff
@@ -71,8 +71,13 @@ BerkeleyEnvironment* GetWalletEnv(const fs::path& wallet_path, std::string& data
+    // Ensure that the directory does not end with a trailing separator to avoid
+    // creating two Berkeley environments in the same directory
+    while ((env_directory.string().back() == '/') || (env_directory.string().back() == '\\'))
+        env_directory = env_directory.remove_trailing_separator();
```
I would prefer the code down there. It might cause an infinity loop on Unix if the wallet filename is `wallet\`
```cpp
while (fs::detail::is_directory_separator(env_directory.string().back())) {
    env_directory.remove_trailing_separator();
}
```

PierreRochard replied:
Thank you, that's much cleaner, fixed!

promag reviewed (commented):
Move the path cleanup here?
https://github.com/bitcoin/bitcoin/blob/c39fa34bfd90b1de0016ad45c207213bd03ecd75/src/wallet/walletutil.cpp#L12

fanquake commented on src/wallet/test/db_tests.cpp:
```diff
@@ -0,0 +1,43 @@
+ */
+BOOST_AUTO_TEST_CASE(get_wallet_env_trailing_slash)
+{
+    const char kPathSeparator =
```
@promag Not sure if you're aware, but anything `boost::filesystem::` should use `fs::` as is already being done here. See https://github.com/bitcoin/bitcoin/blob/master/src/fs.h.

promag replied:
Yes, just wanted give full typr for reference.

PierreRochard replied:
Thank you, that's great, fixed!

ken2812221 commented:
> I'm unable to reproduce the CI failures locally for debugging and the logs don't seem to provide enough information to pinpoint the problem

You can take a look at appveyor CI. That might be helpful.

PierreRochard commented on src/wallet/test/db_tests.cpp:
```diff
@@ -0,0 +1,43 @@
+
+    std::vector<fs::path> wallet_test_paths;
+    std::string trailing_slashes;
+    for (int i = 0; i < 4; ++i ) {
```
Fixed

PierreRochard commented:
@promag I moved the modification to `walletutil.cpp`, excellent suggestion, thank you! 

Force push diff: https://github.com/PierreRochard/bitcoin/commit/5a28a99d2887be85d02cb9a9a062f6bde96f56a2

promag commented on src/wallet/walletutil.cpp:
```diff
@@ -23,5 +23,11 @@ fs::path GetWalletDir()
 
+    // Ensure that the directory does not end with a trailing separator to avoid
+    // creating two Berkeley environments in the same directory
+    while (fs::detail::is_directory_separator(path.string().back())) {
```
I think you could add above:
```cpp
if (gArgs.IsArgSet("-walletdir")) {
    path = gArgs.GetArg("-walletdir", "");
    boost::system::error_code error;
    path = fs::canonical(path, error);
    if (error || !fs::is_directory(path)) {
        ...
```
`fs::canonical` checks if it exists and also cleans the path. For instance, locally calling `canonical("/Users/promag/.///.//")` gives `/Users/promag`.

PierreRochard replied:
Ah `fs::canonical` is even better! What do you think about me adding this to `WalletInit::Verify` [here](https://github.com/bitcoin/bitcoin/blob/0.17/src/wallet/init.cpp#L180-L189) and then `gArgs.ForceSetArg` with the new clean path? 
That way we keep the error checking in one place and the path gets cleaned up once, rather than every time `GetWalletDir` is called.

promag replied:
Sounds good, you could replace `fs::exists` with `fs::canonical` in https://github.com/bitcoin/bitcoin/blob/465a583f9dcc2383eb20bbb98bffa66fe82f2cde/src/wallet/init.cpp#L182

PierreRochard replied:
The nuance with that approach is that we want to avoid transforming a relative file path to an absolute one, per @ryanofsky's commit message there: 

> Specifying paths relative to the current working directory in a daemon process can be dangerous, because files can fail to be located even if the configuration doesn't change, but the daemon is started up differently.

I'm thinking to first check `wallet_dir.is_absolute()` and then use `fs::canonical`

promag commented on src/wallet/test/walletutil_tests.cpp:
```diff
@@ -0,0 +1,28 @@
+{
+    fs::path expected_wallet_dir = SetDataDir("get_wallet_dir");
+    std::string trailing_separators;
+    for (int i = 0; i < 4; ++i) {
```
I would add a couple of cases instead.

tungnguen554 commented:
bitcoin is my first time now


чт, 6 сент. 2018 г. в 21:37, João Barbosa <notifications@github.com>:

> *@promag* commented on this pull request.
> ------------------------------
>
> In src/wallet/walletutil.cpp
> <https://github.com/bitcoin/bitcoin/pull/14146#discussion_r215649285>:
>
> > @@ -23,5 +23,11 @@ fs::path GetWalletDir()
>          }
>      }
>
> +    // Ensure that the directory does not end with a trailing separator to avoid
> +    // creating two Berkeley environments in the same directory
> +    while (fs::detail::is_directory_separator(path.string().back())) {
>
> I think you could add above:
>
> if (gArgs.IsArgSet("-walletdir")) {
>     path = gArgs.GetArg("-walletdir", "");
>     boost::system::error_code error;
>     path = fs::canonical(path, error);
>     if (error || !fs::is_directory(path)) {
>         ...
>
> fs::canonical checks if it exists and also cleans the path. For instance,
> locally calling canonical("/Users/promag/.///.//") gives /Users/promag.
> ------------------------------
>
> In src/wallet/test/walletutil_tests.cpp
> <https://github.com/bitcoin/bitcoin/pull/14146#discussion_r215649928>:
>
> > +#include <boost/test/unit_test.hpp>
> +
> +#include <test/test_bitcoin.h>
> +#include <wallet/walletutil.h>
> +
> +
> +BOOST_FIXTURE_TEST_SUITE(walletutil_tests, TestingSetup)
> +
> +/** If a user passes in a path with a trailing separator as the walletdir, multiple BerkeleyEnvironments
> + * may be created in the same directory which can lead to data corruption.
> + */
> +BOOST_AUTO_TEST_CASE(get_wallet_dir_trailing_separators)
> +{
> +    fs::path expected_wallet_dir = SetDataDir("get_wallet_dir");
> +    std::string trailing_separators;
> +    for (int i = 0; i < 4; ++i) {
>
> I would add a couple of cases instead.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/14146#pullrequestreview-152960954>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AnrjJ5keW5DN5HT-H_Cis0G2g47MegeRks5uYTMvgaJpZM4WZHpt>
> .
>

PierreRochard committed:
wallet: Add walletdir arg unit tests

PierreRochard committed:
wallet: Remove trailing separators from -walletdir arg

PierreRochard commented:
Took up @promag's excellent suggestion to clean up the path with `fs::canonical`. Added unit tests for the `walletdir` interaction in wallet init. If there are other (cross-platform) test cases you would like to see, I'm happy to add them. The commits have completely changed, no value in a force push diff.

ken2812221 commented on src/wallet/test/init_tests.cpp:
```diff
@@ -0,0 +1,77 @@
+#include <wallet/test/init_test_fixture.h>
+
+#include <wallet/wallet.h>
+#include <wallet/init.cpp>
```
Why this include cpp file?

PierreRochard replied:
The `WalletInit` class was moved from the header file to the cpp file in this pull request https://github.com/bitcoin/bitcoin/pull/12836

Should I move it back or is there a bigger picture architectural issue that I'm not seeing?

promag replied:
Include `src/walletinitinterface.h`?

PierreRochard replied:
🤔 how would I create an instance of the `WalletInit` class to test its implementation of `Verify`?

(I'll take "read this book / article" as an answer if I'm out of my depth here haha)

ken2812221 replied:
The program will find the implementation by C++ virtual method table?

promag replied:
There is already an instance. Just replace the include.

PierreRochard replied:
Ah ok, fixed, I added `init.h` as well as it has the global instance in it

PierreRochard commented:
Appveyor is failing, I rebooted into my Windows environment and see a memory leak is being detected, I'm debugging it now.

```
PS C:\Users\pierre\src\bitcoin\build_msvc\x64\Debug>  .\test_bitcoin.exe --log_level=all --show_progress=1  --run_test=init_walletdir_tests | Clip

*** 1 failure is detected in the test module "Bitcoin Test Suite"
Detected memory leaks!
Dumping objects ->
{29230} normal block at 0x0000019CF6A4A530, 160 bytes long.
 Data: <C : \ U s e r s > 43 00 3A 00 5C 00 55 00 73 00 65 00 72 00 73 00
{29229} normal block at 0x0000019CF6B92A00, 24 bytes long.
 Data: <H^              > 48 5E FC B1 F6 7F 00 00 01 00 00 00 01 00 00 00
{29228} normal block at 0x0000019CF6B7B330, 16 bytes long.
 Data: <                > D0 1A B5 F6 9C 01 00 00 00 00 00 00 00 00 00 00
{29227} normal block at 0x0000019CF6B7A930, 16 bytes long.
 Data: <                > A8 1A B5 F6 9C 01 00 00 00 00 00 00 00 00 00 00
{29226} normal block at 0x0000019CF6B7B560, 16 bytes long.
 Data: <                > 80 1A B5 F6 9C 01 00 00 00 00 00 00 00 00 00 00
{29225} normal block at 0x0000019CF6B51A80, 120 bytes long.
 Data: <`       0       > 60 B5 B7 F6 9C 01 00 00 30 A5 A4 F6 9C 01 00 00
{29224} normal block at 0x0000019CF6B7ACA0, 16 bytes long.
 Data: <  <@            > C8 BC 3C 40 01 00 00 00 00 00 00 00 00 00 00 00
{29223} normal block at 0x0000019CF6B92700, 26 bytes long.
 Data: <boost::filesyste> 62 6F 6F 73 74 3A 3A 66 69 6C 65 73 79 73 74 65
{29222} normal block at 0x0000019CF6B93360, 32 bytes long.
 Data: <boost::filesyste> 62 6F 6F 73 74 3A 3A 66 69 6C 65 73 79 73 74 65
{29221} normal block at 0x0000019CF6B7B290, 16 bytes long.
 Data: <p <@            > 70 BC 3C 40 01 00 00 00 00 00 00 00 00 00 00 00
{28961} normal block at 0x0000019CF6B5A050, 214 bytes long.
 Data: <C : \ U s e r s > 43 00 3A 00 5C 00 55 00 73 00 65 00 72 00 73 00
{28884} normal block at 0x0000019CF6B93420, 24 bytes long.
 Data: < ]              > 10 5D FC B1 F6 7F 00 00 01 00 00 00 01 00 00 00
{28883} normal block at 0x0000019CF6B7B7E0, 16 bytes long.
 Data: < i              > 00 69 B8 F6 9C 01 00 00 00 00 00 00 00 00 00 00
{28882} normal block at 0x0000019CF6B86900, 64 bytes long.
 Data: <        P       > E0 B7 B7 F6 9C 01 00 00 50 A0 B5 F6 9C 01 00 00
{26820} normal block at 0x0000019CF6A716F0, 65536 bytes long.
 Data: < K  H    l   { V> B5 4B 04 BA 48 E5 CE FB D0 6C DE 08 1F 7B 81 56
{26819} normal block at 0x0000019CF4F4D040, 208 bytes long.
 Data: <                > 00 00 00 00 00 00 00 00 F0 16 A7 F6 9C 01 00 00
{26816} normal block at 0x0000019CF6A61280, 144 bytes long.
 Data: <C : \ U s e r s > 43 00 3A 00 5C 00 55 00 73 00 65 00 72 00 73 00
{26815} normal block at 0x0000019CF6B79850, 16 bytes long.
 Data: <X <@            > 58 C3 3C 40 01 00 00 00 00 00 00 00 00 00 00 00
{15322} normal block at 0x0000019CF6AC6A90, 16 bytes long.
 Data: <                > E0 EC 02 F5 9C 01 00 00 00 00 00 00 00 00 00 00
{15321} normal block at 0x0000019CF6AC6C70, 16 bytes long.
 Data: <                > B8 EC 02 F5 9C 01 00 00 00 00 00 00 00 00 00 00
{15320} normal block at 0x0000019CF6B4C290, 56 bytes long.
 Data: <                > 90 C2 B4 F6 9C 01 00 00 90 C2 B4 F6 9C 01 00 00
{15319} normal block at 0x0000019CF502EC60, 176 bytes long.
 Data: <                > 00 00 00 00 00 00 00 00 02 00 00 00 00 00 00 00
{9975} normal block at 0x0000019CF7117070, 1048576 bytes long.
 Data: <    [  Y ( -    > 98 17 F8 16 5B 81 F2 59 D9 28 CE 2D DB FC 9B 02
{9974} normal block at 0x0000019CF4F4C490, 208 bytes long.
 Data: <pp              > 70 70 11 F7 9C 01 00 00 00 00 00 00 00 00 00 00
Object dump complete.
PS C:\Users\pierre\src\bitcoin\build_msvc\x64\Debug>
```

promag commented on src/wallet/test/init_test_fixture.cpp:
```diff
@@ -0,0 +1,35 @@
+    m_walletdir_path_cases["trailing"] = m_datadir / "wallets" / sep;
+    m_walletdir_path_cases["trailing2"] = m_datadir / "wallets" / sep / sep;
+
+
```
Remove 2nd empty line.

PierreRochard replied:
Fixed

PierreRochard commented:
I replaced the `wallet/init.cpp` include and removed extraneous whitespace, force push diff here https://github.com/bitcoin/bitcoin/commit/c8b2076d6206ac8fe349c08a97611189d2e729a8
I'm still working on the Windows memory leak issue.

PierreRochard commented:
I solved the problem on my local Windows machine, the fixture destructor was removing the current working directory - apparently this is ok on posix but not on Windows, force push diff: https://github.com/bitcoin/bitcoin/commit/6e2616deffd815051c1ac2e826bd55850e2898f9

Hopefully appveyor agrees that this was the problem 🤞

ken2812221 commented:
The unit tests is spamming error messages while running test_bitcoin. Would it confuse the user?

PierreRochard commented:
> The unit tests is spamming error messages while running test_bitcoin. Would it confuse the user?

It definitely could, I've actually been researching how to properly mock `CClientUIInterface`. A shortcut would be to create a unit-test version of `noui_connect` but I was struggling on how to capture the output to check it in the unit test. If you have ideas I'm interested, whether they fit into the scope of this PR or not, as I'm sure this isn't the last time I'll be hitting `InitMessage` in a unit test.

ken2812221 commented:
@PierreRochard Haven't tested, but can we use `gArgs.SoftSetBoolArg("-printtoconsole", false);`

PierreRochard commented:
Unfortunately that didn't work. In `noui.cpp` the `noui_ThreadSafeMessageBox` function uses `fprintf` for output

ken2812221 commented:
IMO you could just replace it with `LogPrintf`

PierreRochard commented:
I just looked into opening a PR for a simple replace of `fprintf` with `LogPrintf` but I think it would have to be a larger refactoring, as `stderr` output is necessary in some cases, for example when the debug log file can't be opened:
```
AssertionError: [node 0] Expected message "Error: Could not open debug log file \S+$" does not fully match stderr:
""
```

practicalswift commented on src/wallet/test/init_test_fixture.cpp:
```diff
@@ -0,0 +1,42 @@
+void InitWalletDirTestingSetup::SetWalletDir(const fs::path& walletdir_path)
+{
+    gArgs.ForceSetArg("-walletdir", walletdir_path.string());
+}
```
```
2018-09-22 21:11:48 cpplint(pr=14146): src/wallet/test/init_test_fixture.cpp:42:  Could not find a newline character at the end of the file.  [whitespace/ending_newline] [5]
```

practicalswift replied:
```
2018-09-25 20:31:54 clang(pr=14146): wallet/test/init_test_fixture.cpp:42:2: warning: no newline at end of file [-Wnewline-eof]
```

practicalswift commented on src/wallet/test/init_tests.cpp:
```diff
@@ -0,0 +1,78 @@
+
+#include <init.h>
+#include <walletinitinterface.h>
+#include <wallet/wallet.h>
```
Please sort includes and make the grouping consistent to how it is done in other files.

practicalswift commented on src/wallet/test/init_test_fixture.h:
```diff
@@ -0,0 +1,21 @@
+
+#include <test/test_bitcoin.h>
+
+
```
Nit: Extra newline.

laanwj commented:
utACK 2d471636eb9160ab51b08e491e3f003f57adbc36
thanks for finding and addressing this bug

ryanofsky reviewed (commented):
utACK 2d471636eb9160ab51b08e491e3f003f57adbc36. This fix looks like it does the right thing, but it seems like it would be cleaner and more obvious to do the canonicalization in the `GetWalletDir()` function rather than `WalletInit::Verify()`.

In particular I don't like how `GetWalletDir()` can return the wrong result if it is called before `WalletInit::Verify()` and the right result after. Ideally `GetWalletDir()` would just return the same, canonical result all the time, and I'm curious if canonicalizing in `GetWalletDir()` wouldn't work for some reason.

Another more minor issue predating this PR is that `WalletInit::Verify()` does the `is_absolute` check *after* checking whether the directory exists, and so might access the filesystem with a path relative to the working directory. Since bitcoind is a daemon, it's better if it can be agnostic to the working directory, and only access the filesystem with absolute paths (`AbsPathForConfigVal` util function is handy for this).
